File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/TradeItSdkInstance.java
package it.trade.android.sdk;

import it.trade.android.sdk.exceptions.TradeItKeystoreServiceCreateKeyException;
import it.trade.android.sdk.exceptions.TradeItRetrieveLinkedLoginException;
import it.trade.android.sdk.exceptions.TradeItSDKConfigurationException;
import it.trade.android.sdk.internal.TradeItKeystoreService;
import it.trade.android.sdk.manager.TradeItLinkedBrokerManager;
import it.trade.android.sdk.model.TradeItApiClientParcelable;
import it.trade.android.sdk.model.TradeItLinkedBrokerCache;
import it.trade.model.request.TradeItEnvironment;


public class TradeItSdkInstance {
    private TradeItEnvironment environment;
    private TradeItLinkedBrokerManager linkedBrokerManager;
    private TradeItLinkedBrokerCache linkedBrokerCache;
    private TradeItKeystoreService keyStoreService;

    public TradeItSdkInstance(TradeItConfigurationBuilder configurationBuilder) {
        String baseUrl = configurationBuilder.getBaseUrl();
        if (baseUrl != null && !baseUrl.isEmpty()) {
            configurationBuilder.getEnvironment().setBaseUrl(baseUrl);
        }
        initializeTradeItSdkInstance(configurationBuilder);
    }

    private void initializeTradeItSdkInstance(TradeItConfigurationBuilder configurationBuilder) {
        this.environment = configurationBuilder.getEnvironment();
        this.linkedBrokerCache = new TradeItLinkedBrokerCache(configurationBuilder.getContext());

        try {
            this.keyStoreService = new TradeItKeystoreService(configurationBuilder.getContext());
        } catch (TradeItKeystoreServiceCreateKeyException e) {
            throw new TradeItSDKConfigurationException("Error initializing TradeItKeystoreService: ", e);
        }

        try {
            linkedBrokerManager = new TradeItLinkedBrokerManager(new TradeItApiClientParcelable(configurationBuilder.getApiKey(), environment, configurationBuilder.getRequestInterceptorParcelable()), linkedBrokerCache, keyStoreService, configurationBuilder.isPrefetchBrokerListEnabled());
        } catch (TradeItRetrieveLinkedLoginException e) {
            throw new TradeItSDKConfigurationException("Error initializing TradeItLinkedBrokerManager: ", e);
        }
    }

    public TradeItLinkedBrokerManager getLinkedBrokerManager() {
        return linkedBrokerManager;
    }

    public TradeItEnvironment getEnvironment() {
        return environment;
    }

    public TradeItLinkedBrokerCache getLinkedBrokerCache() {
        return linkedBrokerCache;
    }
}

File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/TradeItSdkInstance.kt
package it.trade.android.sdk

import it.trade.android.sdk.exceptions.TradeItKeystoreServiceCreateKeyException
import it.trade.android.sdk.exceptions.TradeItRetrieveLinkedLoginException
import it.trade.android.sdk.exceptions.TradeItSDKConfigurationException
import it.trade.android.sdk.internal.TradeItKeystoreService
import it.trade.android.sdk.manager.TradeItLinkedBrokerManager
import it.trade.android.sdk.model.TradeItApiClientParcelable
import it.trade.android.sdk.model.TradeItLinkedBrokerCache
import it.trade.model.request.TradeItEnvironment


class TradeItSdkInstance(configurationBuilder: TradeItConfigurationBuilder) {
    var environment: TradeItEnvironment? = null
        private set
    var linkedBrokerManager: TradeItLinkedBrokerManager? = null
        private set
    var linkedBrokerCache: TradeItLinkedBrokerCache? = null
        private set
    private var keyStoreService: TradeItKeystoreService? = null

    init {
        val baseUrl = configurationBuilder.baseUrl
        if (baseUrl != null && !baseUrl.isEmpty()) {
            configurationBuilder.environment.baseUrl = baseUrl
        }
        initializeTradeItSdkInstance(configurationBuilder)
    }

    private fun initializeTradeItSdkInstance(configurationBuilder: TradeItConfigurationBuilder) {
        this.environment = configurationBuilder.environment
        this.linkedBrokerCache = TradeItLinkedBrokerCache(configurationBuilder.context)

        try {
            this.keyStoreService = TradeItKeystoreService(configurationBuilder.context)
        } catch (e: TradeItKeystoreServiceCreateKeyException) {
            throw TradeItSDKConfigurationException("Error initializing TradeItKeystoreService: ", e)
        }

        try {
            linkedBrokerManager = TradeItLinkedBrokerManager(TradeItApiClientParcelable(configurationBuilder.apiKey, environment, configurationBuilder.requestInterceptorParcelable), linkedBrokerCache, keyStoreService, configurationBuilder.isPrefetchBrokerListEnabled)
        } catch (e: TradeItRetrieveLinkedLoginException) {
            throw TradeItSDKConfigurationException("Error initializing TradeItLinkedBrokerManager: ", e)
        }

    }
}
